Layout API groundwork for GeoPDF work

nyalldawson opened the pull request:
This PR splits out the self-contained API changes from #31154 (to avoid a mega PR forming) .

nyalldawson committed:
Add API to register rendered feature handlers for a layout map

nyalldawson committed:
Provide rendered feature handlers access to render context

nyalldawson committed:
Refactor out logic for exporting layouts into separate layers

nyalldawson committed:
Add method to determine whether a layout map has any grids or overviews enabled

nyalldawson committed:
API Improvements to handling layered layout exports, add tests

Fixes #20253

nyalldawson committed:
Also allow retrieving map layer ID associated with a layout export layer

roya0045 commented:
I have a question regarding the signification of this PR. From what I can remember PDF made with Qgis can be imported in Qgis and are properly georeferenced. Whereas GeoPDF seem to be a distinct (proprietary?) format.

Could you explain what this entails and what are the advantages compared to the typical georeferenced PDF that are made?

nyalldawson commented:
@roya45 currently, the pdfs created in qgis can be used as raster datasets. By contrast geospatial pdf/geopdf can be used as a vector format, storing the features and attributes themselves.

It has other benefits too - such as allowing individual layers to be toggled on and off in the pdf reader interactively (which works in almost all pdf readers, eg. Open source ones as well as acrobat reader)

nyalldawson committed:
[layouts] Improve logic of splitting layouts into separate layers when
exporting to a multi-layer format

Now, items are either
- placed onto the same layer as other items (simple items like labels, lines, pictures)
- placed onto the same layer as only other items of equal types (semi-complex items like scalebars or pages)
- placed onto their own unique layers (complex items like legends, maps)

Plus lots of tests covering this, where previously there was few

nyalldawson committed:
When an item is isolated on its own layer, use the item's id as the layer name

nyalldawson committed:
When multiple items of same type are on a layer, use type name as layer name

nyalldawson committed:
[layouts] Don't export empty layers in layered exports

nyalldawson committed:
Add API to set/retrieve plural variants of layout item type names (e.g. Labels instead of Label)

nyalldawson committed:
Much improved, more descriptive names for exported layers

E.g. use lists of item types to describe multi-item layers

nyalldawson committed:
Better display names for page items

These can't have ids set, so default to a nicer looking name

nyalldawson committed:
Another tweak to export names: indicate source map item in map item sublayer names

nirvn commented on src/core/qgsrenderedfeaturehandlerinterface.h:
```diff
@@ -50,10 +50,19 @@ class CORE_EXPORT QgsRenderedFeatureHandlerInterface
+
+      /**
+       * Constructor for RenderedFeatureContext.
+       * \param renderContext The render context which was used while rendering feature.
```
I assume the pointer here isn't made to have its ownership transferred?

nyalldawson replied:
I swapped it to a const reference for clarity

nirvn commented on tests/src/core/testqgslayoutexporter.cpp:
```diff
@@ -0,0 +1,260 @@
+  res = exporter.handleLayeredExport( items, exportFunc );
+  QCOMPARE( res, QgsLayoutExporter::Success );
+  QCOMPARE( layerIds, QList< unsigned int >() << 1 << 2 );
+  QCOMPARE( layerNames, QStringList() << QStringLiteral( "Pages" ) << QStringLiteral( "<Label>" ) );
```
Is there a reason you enclose Label within lesser than / greater than signs, and not do so for Page(s)? I'd remove the symbols here (you also don't do it when naming layers containing multiple item types, e.g. "Label, Shape").

nyalldawson replied:
It's coming straight from the layout item "displayName" (i.e. the value shown in the layout items tree widget). By default this is of the format <Scalebar>, <Legend>, etc, UNLESS the item has an ID set, in which case it's the ID. Also, some item types override this, e.g. labels, which return just the first portion of the label's text.

So while in principal I agree that the mix looks messy, I can't think of a clean way to avoid this.

nirvn replied:
Could you clean it up in the nameForLayerWithItems() function, whereas you could strip the first and last chars of a single item layer's displayName of those are < and > ?

nirvn commented on tests/src/core/testqgslayoutexporter.cpp:
```diff
@@ -0,0 +1,260 @@
+  res = exporter.handleLayeredExport( items, exportFunc );
+  QCOMPARE( res, QgsLayoutExporter::Success );
+  QCOMPARE( layerIds, QList< unsigned int >() << 1 << 2 << 3 << 4 << 5 );
+  QCOMPARE( layerNames, QStringList() << QStringLiteral( "Pages" ) << QStringLiteral( "Labels, Shape" ) << QStringLiteral( "<Scalebar>" ) << QStringLiteral( "<Label>" ) << QStringLiteral( "Scalebars" ) );
```
<> vs non <>.

nirvn commented on tests/src/core/testqgslayoutexporter.cpp:
```diff
@@ -0,0 +1,260 @@
+  QVERIFY( pointsLayer->isValid() );
+
+  p.addMapLayer( linesLayer );
+  p.addMapLayer( pointsLayer );
```
You are not using the pointsLayer anywhere after adding it to the project, did you mean to test multiple exported layer name IDs somewhere?

nyalldawson replied:
Just a copy/paste error, I'll remove. Handling of multiple map layer's is tested in depth in testqgslayoutitemmap.cpp

nirvn commented on src/core/layout/qgslayoutitem.h:
```diff
@@ -414,16 +414,60 @@ class CORE_EXPORT QgsLayoutItem : public QgsLayoutObject, public QGraphicsRectIt
+    {
+      CanGroupWithAnyOtherItem, //!< Item can be placed on a layer with any other item (default behavior)
+      CanGroupWithItemsOfSameType, //!< Item can only be placed on layers with other items of the same type, but multiple items of this type can be grouped together
+      MustPlaceInOwnLayer, //!< Item must be placed in its own individual layer
```
Should you edit the export layer behavior of 3D map items to be set to MustPlaceInOwnLayer ?

nyalldawson replied:
I don't think so -- these become images in the exports, so are easy to manage when in the same layer as other simple item types. I want to keep MustPlaceInOwnLayer only for complex items which have lots of parts in vector exports (like legends), and would be tricky to manage if they were mixed in a layer with other items.

nirvn replied:
K; I was wondering whether it was worth harmonizing 3D vs 2D map items, i.e. if 2D is a layer of its own, 3D should follow that. But there was indeed no reason other than that. No strong feeling about this, perfectly fine for it to stay as is.

nirvn commented on src/core/layout/qgslayoutexporter.cpp:
```diff
@@ -1466,6 +1436,169 @@ bool QgsLayoutExporter::georeferenceOutputPrivate( const QString &file, QgsLayou
+  if ( items.count() == 1 )
+  {
+    if ( QgsLayoutItem *layoutItem = dynamic_cast<QgsLayoutItem *>( items.at( 0 ) ) )
+      return layoutItem->displayName();
```
I commented in your added tests on the less-than-ideal mixture of layer names using lesser than / greater than symbols vs. names not surrounded by those symbols. E.g. Page, Labels, scalebar, Map 1: lines, <Label>. IMO, it'd be worth fixing this if possible.

nirvn commented on src/core/layout/qgslayoutexporter.cpp:
```diff
@@ -1466,6 +1436,169 @@ bool QgsLayoutExporter::georeferenceOutputPrivate( const QString &file, QgsLayou
   return true;
 }
 
+QString nameForLayerWithItems( const QList< QGraphicsItem * > items, unsigned int layerId )
```
Any reason why this function is not part of the QgsLayoutExporter class?

nyalldawson replied:
It's just a little implementation detail -- I don't see any need to expose it to the class.

nirvn reviewed (commented):
@nyalldawson , looked at the code, looking good (as always). A few comments added.

nyalldawson committed:
Remove unused layer

nyalldawson committed:
Use reference instead of pointer

nyalldawson committed:
Cleanup default item displayName formatting in export layers

nirvn commented on src/core/layout/qgslayoutexporter.cpp:
```diff
@@ -1441,7 +1441,13 @@ QString nameForLayerWithItems( const QList< QGraphicsItem * > items, unsigned in
+      QString name = layoutItem->displayName();
+      // cleanup default item ID format
+      if ( name.startsWith( '<' ) && name.endsWith( '>' ) )
+        name = name.mid( 1, name.length() - 2 );
```
@nyalldawson , :tada:
